teps/0145-cel-in-whenexpression.md

```diff
@@ -0,0 +1,375 @@
+```
+
+With CEL we can use a more intuitive expression to support it.
+```yaml
+when:
+  cel: "'$(params.param1)' != '' || '$(params.param2)' == '8.5' || $(params.param2)' == '8.6'"
+```
+
+#### Use CEL to Support Numeric Comparisons
+
+This is not possible with current `WhenExpression`. With CEL we can do:
+
+```yaml
+  when:
+    - cel: "'$(tasks.unit-test.results.test-coverage)' >= 0.9"
+```
+
+[Note:][numeric note] currently there are no automatic arithmetic conversions for the numeric types (int, uint, and double), so the emitted results must be the same numeric type to make sure the evaluation works.
+
+#### Use CEL to Support Pattern Matching:
+
+This is not possible with current `WhenExpression`. With CEL we can do:
+
+```yaml
+  when:
+    - cel: '$(params.branch)'.matches('release/.*')"
+```
+
+### Variables Substitution
+
+The CEL in When Expressions should support current Tekton’s Params and Results [string substitutions](https://github.com/tektoncd/pipeline/blob/main/docs/variables.md#variables-available-in-a-pipeline). Array and object substitutions are not supported because Tekton doesn’t support whole array/object replacements in a string.
+
+We should support Tekton’s variable substitution for 2 reasons:
+- This is consistent with all other variable substitution syntax in Tekton, and thus more intuitive for Tekton users;
+- This would help to build the dag of Pipeline Tasks, as tasks in a pipeline will check the result reference and build the dependency between Tasks.
+
+We cannot use the current Tekton's variable substitution directly. Because the current variables substitution has the the similar security concern as [TEP-0146](https://github.com/tektoncd/community/pull/1077), for example, users can inject a CEL expression from the `Param` and it could alter the result of the CEL which referenced that `Param` in `WhenExpression`, or execute a CEL which consume more resources than expected.
+
+To address this, we need to:
+- Prevent the attack by adding sanity check;
+- Explicitly document that passing CEL expression from param is not allowed and won’t be executed.
+
+The solution we proposed is to let CEL handle the variable substitution:
+1. Add params, results, context variables to CEL's environment, similar like [Tekton Triggers](https://github.com/tektoncd/triggers/blob/main/pkg/interceptors/cel/cel.go#L104C1-L112)
+2. Find all variable references in the CEL expression string and remove the '$()' from those references.
+3. CEL would do the replacement with 1 and 2 with built-in variable substitution
+
+
+### Validation
+
+The new fields will be gated by `cel-in-whenexpression`, which will default to false while the feature is in alpha.
+
+The validation webhook will validate
+1. The feature flag is enabled
+2. If the flag is enabled, either the current when input+operator+values or cel is used, users cannot use both at the same time for one WhenExpression, but they can use different syntax for different WhenExpressions.
```

Comment:
This might have already been mentioned during the WG, just wondering if this would need to include the backwards compatibility?

Reply:
Sure, will add now